### PR TITLE
feature/add presets for more clang compilers

### PR DESCRIPTION
**Pull request by anders-wind**
### Why
We wish to support - and show that we do - newer compilers.

### What
- Added clang-15 and clang-17 configurations and test them in CI. 
- Default to clang-17 for dev builds, clang-format and clang-tidy

### Notes
Note that clang-16 have a bug and stronk is not compatible with this version: 
See https://github.com/llvm/llvm-project/issues/62518 and https://github.com/llvm/llvm-project/issues/62344 for the clang-16 frontend crashes.

### CODE REVIEW

**Comment by anders-wind** [comment]
Still works with `Ubuntu clang version 17.0.0 (++20230822075756+5bd2c0c745d4-1~exp1~20230822075815.29)`. Looks like we can go on with the clang-17 upgrade

**Comment by anders-wind** [comment]
We drop supporting clang-16 but clang-17 seem to be working

**Review by mikaellindemann — COMMENTED** [review]
Any particular reason for not trying to run with clang-17 in CI?

**Comment by anders-wind** [comment]
> Any particular reason for not trying to run with clang-17 in CI?
now we do :)

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -16,13 +16,13 @@ on:
   workflow_dispatch:
 
 env:
-  VCPKG_COMMIT: "6f7ffeb18f99796233b958aaaf14ec7bd4fb64b2"
+  VCPKG_COMMIT: "c8696863d371ab7f46e213d8f5ca923c4aef2a00"
```
Comment (mikaellindemann) [inline .github/workflows/ci.yml:19]: You wrote in one of the commits that this was the 2024.01 version, but it seems to be the hash of the [2023.12](https://github.com/microsoft/vcpkg/releases/tag/2023.12.12) version

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -182,15 +208,32 @@ jobs:
...
+        uses: KyleMayes/install-llvm-action@v1
+        with:
+          version: "17.0"
+      
+      - name: Fix clang installations
+        if: matrix.os == 'ubuntu-22.04'
+        run: |
+          cp ./llvm/bin/clang++ ./llvm/bin/clang++-17
+          cp ./llvm/bin/clang-format ./llvm/bin/clang-format-17
+          cp ./llvm/bin/clang-tidy ./llvm/bin/clang-tidy-17
```
Comment (mikaellindemann) [inline .github/workflows/ci.yml:236]: Wouldn't these only be necessary for the ubuntu-22.04, clang-17 run?

I'm unsure how the matrix works, but would suspect it to run these steps per compiler with the current setup.

**Review by mikaellindemann — APPROVED** [review]

File: .github/workflows/ci.yml (hunk 1 above)
Comment (anders-wind) [reply .github/workflows/ci.yml:19]: I was referring to vcpkg not abseil in the commit message

Comment (mikaellindemann) [reply .github/workflows/ci.yml:19]: But this is also the vcpkg commit hash pointing to 2023.12?

Comment (anders-wind) [reply .github/workflows/ci.yml:19]: Ahh yea my bad! I was just looking at our main repo but I missed that it was a revert commit. But in any case this matches that version :) still fine

File: .github/workflows/ci.yml (hunk 2 above)
Comment (anders-wind) [reply .github/workflows/ci.yml:236]: The gcc ones are also running clang-tidy

Comment (anders-wind) [reply .github/workflows/ci.yml:236]: But I agree this is not really optimal

**Pull request merged**